File: src/main/java/fr/free/nrw/jakaroma/Jakaroma.java
package fr.free.nrw.jakaroma;

import com.atilika.kuromoji.ipadic.Token;
import com.atilika.kuromoji.ipadic.Tokenizer;

import java.util.List;
import java.util.Scanner;

public class Jakaroma {

    private static final boolean DEBUG = false;
    private boolean isPronunciation = false;

    /**
     * for those who use it from the command line
     */
    public static void main(String... args) {
        String input;
        if (args.length == 0 || args[0].isEmpty()) {
            System.err.println("Jakaroma warn: no argument passed, reading stdin...");
            System.err.flush();

            try (Scanner scanner = new Scanner(System.in).useDelimiter("\\A")) {
                input = (scanner.hasNext() ? scanner.next() : "");
            }
        } else {
            input = args[0];
        }

        Jakaroma instance = new Jakaroma();
        instance.isPronunciation = true;
        instance.convert(input, true, true);
    }

    /**
     * Converts kanji, etc to romaji
     *
     * @param input           the string to convert
     * @param trailingSpace   add trailing space after the token
     * @param capitalizeWords start token translation with a capital letter
     * @return romaji representation
     */
    public String convert(String input, boolean trailingSpace, boolean capitalizeWords) {
        Tokenizer tokenizer = new Tokenizer();
        List<Token> tokens = tokenizer.tokenize(halfWidthToFullWidth(input));
        int tokensSize = tokens.size();
        StringBuilder buffer = new StringBuilder();
        KanaToRomaji kanaToRomaji = new KanaToRomaji();

        // Display tokens
        if (DEBUG) {
            tokens.forEach(token -> System.out.println(token.getSurface() + "\t" + token.getAllFeatures()));
        }

        for (int i = 0; i < tokensSize; i++) {
            Token token = tokens.get(i);
            String type = token.getAllFeaturesArray()[1];
            if (DEBUG) {
                System.out.println("Token: " + token.getSurface());
                System.out.println("Type: " + type);
            }
            // keep newlines unaltered
            if (token.getAllFeaturesArray()[0].equals("記号")) {
                buffer.append(kanaToRomaji.convert(token.getSurface())); // Append surface form
                continue;
            }

            switch (type) {
                case "数": // Example: 4
                case "アルファベット": // Example: ｂ (double-width alphabet)
                case "サ変接続": // Example: , (connection symbols)
                    buffer.append(kanaToRomaji.convert(token.getSurface()));  // TODO - exception list for when this false positives
                    continue; // avoid extra whitespace after symbols
                default:
                    String romaji;
                    // Kuromoji provided no katakana?
                    if (getKatakana(token).equals("*")) {
                        // Look up all token characters against keys in Katakana map
                        // token may be composed of katakana but tripped up kuromoji
                        // Deal with small tsu
                        romaji = kuromojiFailedConvert(token, kanaToRomaji);
                    } else {  // kanji has been converted to katakana
                        // Convert katakana to romaji
                        String katakana = getKatakana(token);

                        if (DEBUG) {
                            System.out.println("Katakana: " + katakana);
                        }

                        String currentRomaji = kanaToRomaji.convert(katakana);
                        // sokuon at end of token list: exclamation mark
                        if (katakana.endsWith("ッ") && i == tokensSize - 1) {
                            int lastIndex = katakana.length() - 1;
                            katakana = katakana.substring(0, lastIndex);
                            // Remove space added in last loop
                            if (i != 0) {
                                buffer.deleteCharAt(buffer.length() - 1);
                            }
                            romaji = kanaToRomaji.convert(katakana) + "!";

                            if (DEBUG) {
                                System.out.println("Exclamation sokuon, romaji becomes: " + romaji);
                            }
                        }
                        // sokuon between tokens, merge with next token
                        else if (katakana.endsWith("ッ")) {
                            romaji = smallTsuRomaji(token, tokens.get(i + 1), kanaToRomaji);
                            if (DEBUG) {
                                System.out.println("Sokuon detected, merged romaji :" + romaji);
                            }
                            // Skip next token since it has been processed here and merged
                            i++;
                        } else {
                            romaji = currentRomaji;
                        }
                    }
                    // Capitalization
                    if (!romaji.isEmpty()) {
                        if (capitalizeWords) {
                            buffer.append(romaji.substring(0, 1).toUpperCase());
                            buffer.append(romaji.substring(1));
                        } else {
                            // Convert foreign katakana words to uppercase
                            if (token.getSurface().equals(token.getReading())) { // detect katakana
                                romaji = romaji.toUpperCase();
                            }
                            buffer.append(romaji);
                        }
                    }
            }
            if (trailingSpace && i != tokensSize - 1) {
                buffer.append(" ");
            }
        }
        System.out.println(buffer);
        return buffer.toString();
    }

    /*
     * @param token
     */
    private String getKatakana(Token token) {
        String katakana;
        if (isPronunciation) {
            katakana = token.getPronunciation();
        } else {
            katakana = token.getReading();
        }
        return katakana;
    }

    private String smallTsuRomaji(Token token, Token nextToken, KanaToRomaji kanaToRomaji) {
        String romaji;
        String nextRomaji = kanaToRomaji.convert(getKatakana(nextToken));
        String currentRomaji = kanaToRomaji.convert(getKatakana(token).substring(0, token.getSurface().length() - 1));
        romaji = currentRomaji + nextRomaji.charAt(0) + nextRomaji;
        return romaji;
    }

    private String kuromojiFailedConvert(Token token, KanaToRomaji kanaToRomaji) {
        StringBuilder buffer = new StringBuilder();
        String surface = token.getSurface();

        if (surface.contains("ッ") && !surface.endsWith("ッ")) {
            String[] splitTokenSurface = surface.split("ッ");
            String romaji1 = kanaToRomaji.convert(splitTokenSurface[0]);
            String romaji2 = kanaToRomaji.convert(splitTokenSurface[1]);
            buffer.append(romaji1).append(romaji2.charAt(0)).append(romaji2);
        } else {
            for (char c : token.getSurface().toCharArray()) {
                // Katakana character? Convert to romaji
                if (KanaToRomaji.getDictionary().containsKey(String.valueOf(c))) {
                    buffer.append(kanaToRomaji.convert(String.valueOf(c)));
                } else {
                    // Append as is
                    buffer.append(c);
                }
            }
        }
        return buffer.toString();
    }

    /**
     * Full-angle string conversion half-corner string
     * 1, half-width characters are starting from 33 to 126 end
     * 2, the full-width character corresponding to the half-width character is from 65281 start to 65374 end
     * 3, the half corner of the space is 32. The corresponding Full-width space is 12288
     * The relationship between Half-width and Full-width is obvious, except that the character offset is 65248 (65281-33 = 65248).
     *
     * @param fullWidthStr Non-empty full-width string
     * @return Half-angle string
     */
    private String halfWidthToFullWidth(String fullWidthStr) {
        if (null == fullWidthStr || fullWidthStr.length() <= 0) {
            return "";
        }
        char[] arr = fullWidthStr.toCharArray();
        for (int i = 0; i < arr.length; ++i) {
            int charValue = arr[i];
            if (charValue >= 65281 && charValue <= 65374) {
                arr[i] = (char) (charValue - 65248);
            } else if (charValue == 12288) {
                arr[i] = (char) 32;
            }
        }
        return new String(arr);
    }
}

File: src/main/kotlin/fr/free/nrw/jakaroma/Jakaroma.kt
package fr.free.nrw.jakaroma

import com.atilika.kuromoji.ipadic.Token
import com.atilika.kuromoji.ipadic.Tokenizer

private const val DEBUG = false

class Jakaroma {
    private var isPronunciation = false // Class property
    private val tokenizer = Tokenizer()

    /**
     * Converts kanji, etc to romaji
     *
     * @param input           the string to convert
     * @param trailingSpace   add trailing space after the token
     * @param capitalizeWords start token translation with a capital letter
     * @return romaji representation
     */
    fun convert(input: String, trailingSpace: Boolean, capitalizeWords: Boolean): String {
        val tokens = tokenizer.tokenize(halfWidthToFullWidth(input))
        val tokensSize = tokens.size
        val buffer = StringBuilder()

        if (DEBUG) {
            tokens.forEach { token -> println("${token.surface}\t${token.allFeatures}") }
        }

        var i = 0 // Initialize index for while loop
        while (i < tokensSize) {
            val token = tokens[i]
            val type = token.allFeaturesArray[1] // Safe if allFeaturesArray always has at least 2 elements

            if (DEBUG) {
                println("Token: ${token.surface}")
                println("Type: $type")
            }

            if (token.allFeaturesArray[0] == "記号") { // Safe if allFeaturesArray always has at least 1 element
                buffer.append(KanaToRomaji.convert(token.surface))
                i++ // Increment index
                continue
            }

            when (type) {
                "数", "アルファベット", "サ変接続" -> {
                    buffer.append(KanaToRomaji.convert(token.surface))
                }
                else -> {
                    var romaji: String
                    val katakanaReading = getKatakana(token)

                    if (katakanaReading == "*") {
                        romaji = kuromojiFailedConvert(token)
                    } else {
                        val currentKatakana = katakanaReading
                        if (DEBUG) {
                            println("Katakana: $currentKatakana")
                        }

                        if (currentKatakana.endsWith("ッ")) {
                            if (i == tokensSize - 1) { // Sokuon at the very end
                                romaji = if (currentKatakana.length > 1) {
                                    // Remove potential preceding space if this is the first actual output character.
                                    // The original Java code had complex space deletion logic here.
                                    // A simple approach: if buffer ends with space, remove it.
                                    if (buffer.isNotEmpty() && buffer.last() == ' ') {
                                        buffer.deleteCharAt(buffer.length - 1)
                                    }
                                    KanaToRomaji.convert(currentKatakana.substring(0, currentKatakana.length - 1)) + "!"
                                } else {
                                    "!" // Just "ッ" at the end becomes "!"
                                }
                                if (DEBUG) println("Exclamation sokuon, romaji becomes: $romaji")
                            } else { // Sokuon followed by another token
                                romaji = smallTsuRomaji(token, tokens[i + 1])
                                if (DEBUG) println("Sokuon detected, merged romaji: $romaji")
                                i++ // Crucial: skip the next token as it has been processed
                            }
                        } else {
                            romaji = KanaToRomaji.convert(currentKatakana)
                        }
                    }

                    if (romaji.isNotEmpty()) {
                        if (capitalizeWords) {
                            buffer.append(romaji.first().uppercaseChar()).append(romaji.substring(1))
                        } else {
                            if (token.surface == katakanaReading && isAllKatakana(token.surface)) {
                                buffer.append(romaji.uppercase())
                            } else {
                                buffer.append(romaji)
                            }
                        }
                    }
                }
            }

            // Add trailing space if needed
            // Only add space if something was appended in this iteration and it's not the last token
            if (trailingSpace && i < tokensSize - 1) {
                // Check if the current token actually produced output
                // This is a bit tricky as the buffer might have been modified by previous tokens.
                // A simpler rule: if buffer is not empty and does not end with a space, add one.
                // This might differ slightly from original's complex space logic.
                if (buffer.isNotEmpty() && buffer.last() != ' ') {
                    // Specific types in original code avoided trailing spaces by `continue`
                    // Here, we ensure those types don't add a space if they were the last thing processed.
                    val isSymbolLike = type in listOf("数", "アルファベット", "サ変接続", "記号")
                    if (!isSymbolLike || KanaToRomaji.convert(token.surface).isNotEmpty()) { // Add space for symbols only if they produced output
                        buffer.append(" ")
                    }
                }
            }
            i++
        }

        // Final trim of trailing space if any exists
        if (buffer.isNotEmpty() && buffer.last() == ' ') {
            buffer.deleteCharAt(buffer.length - 1)
        }

        if (DEBUG) {
            println(buffer.toString())
        }
        return buffer.toString()
    }

    private fun getKatakana(token: Token): String {
        val result = if (isPronunciation) {
            token.pronunciation
        } else {
            token.reading
        }
        return result ?: "*"
    }

    private fun smallTsuRomaji(currentToken: Token, nextToken: Token): String {
        val currentKatakana = getKatakana(currentToken) // e.g., "カッ"
        val nextKatakana = getKatakana(nextToken)     // e.g., "ト"

        // Convert current token's katakana part (before "ッ") to romaji
        val currentRomajiPart = KanaToRomaji.convert(currentKatakana.dropLast(1)) // "カ" -> "ka"

        val nextTokenRomaji = KanaToRomaji.convert(nextKatakana) // "ト" -> "to"

        return if (nextTokenRomaji.isNotEmpty()) {
            "$currentRomajiPart${nextTokenRomaji.first()}$nextTokenRomaji" // "ka" + "t" + "to" = "katto"
        } else {
            currentRomajiPart // Should not happen if nextToken is valid
        }
    }

    private fun kuromojiFailedConvert(token: Token): String {
        val surface = token.surface
        val buffer = StringBuilder()
        var idx = 0
        while (idx < surface.length) {
            val char = surface[idx]
            if (char == 'ッ') {
                if (idx + 1 < surface.length) {
                    val nextChar = surface[idx + 1]
                    val nextRomaji = KanaToRomaji.convert(nextChar.toString())
                    if (nextRomaji.isNotEmpty() && nextRomaji != nextChar.toString()) {
                        // If next char is a convertible Kana, double its first consonant
                        buffer.append(nextRomaji.first())
                    } else {
                        // If next char is not Kana or some other symbol, append 't' or 'tsu' for 'ッ'
                        buffer.append('t') // Or "tsu" or "" depending on desired behavior
                    }
                } else {
                    // Small tsu at the end of a surface form when Kuromoji fails to provide reading
                    buffer.append('!') // Or "tsu" or ""
                }
            } else {
                buffer.append(KanaToRomaji.convert(char.toString()))
            }
            idx++
        }
        return buffer.toString()
    }

    private fun isAllKatakana(text: String): Boolean {
        if (text.isEmpty()) return false
        return text.all { char ->
            // Check for Katakana Unicode block or Prolonged Sound Mark
            (char in '\u30A0'..'\u30FF') || char == '\u30FC'
        }
    }

    private fun halfWidthToFullWidth(fullWidthStr: String?): String {
        if (fullWidthStr.isNullOrEmpty()) {
            return ""
        }
        return fullWidthStr.map { char ->
            when (char.code) {
                // Full-width ASCII variants to Half-width ASCII
                in 0xFF01..0xFF5E -> (char.code - 0xFEE0).toChar()
                // Full-width space (Ideographic Space) to standard space
                0x3000 -> ' '
                else -> char
            }
        }.joinToString("")
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val instance = Jakaroma()
            instance.isPronunciation = true // Default for CLI tool

            val input: String = if (args.isEmpty() || args[0].isBlank()) {
                System.err.println("Jakaroma warn: no argument passed, reading stdin...")
                System.err.flush()
                // Read all input from stdin, trim whitespace
                System.`in`.bufferedReader().readText().trim()
            } else {
                args[0]
            }

            if (input.isNotEmpty()) {
                val result = instance.convert(input, trailingSpace = true, capitalizeWords = true)
                println(result)
            } else {
                 System.err.println("Jakaroma warn: input is empty, nothing to convert.")
            }
        }
    }
}
